precompiles/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+
+impl<Runtime> Precompile for CrowdloanRewardsWrapper<Runtime>
+where
+	Runtime: pallet_crowdloan_rewards::Config + pallet_evm::Config,
+	Runtime::AccountId: From<H160>,
+	BalanceOf<Runtime>: TryFrom<U256> + Debug,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	Runtime::Call: From<pallet_crowdloan_rewards::Call<Runtime>>,
+{
+	fn execute(
+		input: &[u8], //Reminder this is big-endian
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<PrecompileOutput, ExitError> {
+		log::trace!(target: "crowdloan-rewards-precompile", "In crowdloan rewards wrapper");
+
+		// Basic sanity checking for length
+		// https://solidity-by-example.org/primitives/
+
+		const SELECTOR_SIZE_BYTES: usize = 4;
+
+		if input.len() < 4 {
+			return Err(ExitError::Other("input length less than 4 bytes".into()));
+		}
+
+		log::trace!(target: "crowdloan-rewards-precompile", "Made it past preliminary length check");
+		log::trace!(target: "crowdloan-rewards-precompile", "context.caller is {:?}", context.caller);
+
+		// Parse the function selector
+		// These are the four-byte function selectors calculated from the CrowdloanInterface.sol
+		// according to the solidity specification
+		// https://docs.soliditylang.org/en/v0.8.0/abi-spec.html#function-selector
+		let inner_call = match input[0..SELECTOR_SIZE_BYTES] {
+			// Check for accessor methods first. These return results immediately
+			[0x53, 0x44, 0x0c, 0x90] => {
+				return Self::is_contributor(&input[SELECTOR_SIZE_BYTES..], target_gas);
+			}
+			[0x76, 0xf7, 0x02, 0x49] => {
+				return Self::reward_info(&input[SELECTOR_SIZE_BYTES..], target_gas);
+			}
+			[0x4e, 0x71, 0xd9, 0x2d] => Self::claim()?,
+
+			[0xaa, 0xac, 0x61, 0xd6] => Self::update_reward_address(&input[SELECTOR_SIZE_BYTES..])?,
+			_ => {
+				log::trace!(
+					target: "crowdloan-rewards-precompile",
+					"Failed to match function selector in crowdloan rewards precompile"
+				);
+				return Err(ExitError::Other(
+					"No crowdloan rewards wrapper method at given selector".into(),
+				));
+			}
+		};
+
+		let outer_call: Runtime::Call = inner_call.into();
+		let info = outer_call.get_dispatch_info();
+
+		// Make sure enough gas
+		if let Some(gas_limit) = target_gas {
+			let required_gas = Runtime::GasWeightMapping::weight_to_gas(info.weight);
+			if required_gas > gas_limit {
+				return Err(ExitError::OutOfGas);
+			}
+		}
+		log::trace!(target: "crowdloan-rewards-precompile", "Made it past gas check");
+
+		// Dispatch that call
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+
+		log::trace!(target: "crowdloan-rewards-precompile", "Gonna call with origin {:?}", origin);
+
+		match outer_call.dispatch(Some(origin).into()) {
+			Ok(post_info) => {
+				let gas_used = Runtime::GasWeightMapping::weight_to_gas(
+					post_info.actual_weight.unwrap_or(info.weight),
+				);
```

Comment:
Is it possible that `gas_used` here could be > gas provided? Or is this guaranteed to be the same as `required_gas` above?

Reply:
I have been checking that substrate uses the get_dispatch_info to calculate weights in pallets like utility and others. So I assume they should coincide